### PR TITLE
Support wells in sequential poromechanics + 3 other PRs

### PR DESCRIPTION
**Pull request by paveltomin**
Make wells compatible with poromechanics, example for single-phase case:
image

Mechanical initialization would not work when mechanics-only solver is used because bulkDensity is 0, but works ok if coupled solver is applied (documented here https://github.com/GEOS-DEV/GEOS/issues/2698)

Rebase: https://github.com/GEOS-DEV/integratedTests/pull/59

Merged with https://github.com/GEOS-DEV/GEOS/pull/2766, https://github.com/GEOS-DEV/GEOS/pull/2687, and https://github.com/GEOS-DEV/GEOS/pull/2754 to speedup merging

### CODE REVIEW

File: inputFiles/poromechanics/PoroElastic_staircase_singlephase_3d_seq.xml
Comment (joshua-white) [inline inputFiles/poromechanics/PoroElastic_staircase_singlephase_3d_seq.xml:13]: I know you are just testing right now, but we should eventually use the xml include functionality to have a `PoroElastic_staircase_singlephase_3d_base.xml` and then two concise downstream files, `_fim` and `_seq`.  We are only really changing two lines.

**Comment by paveltomin** [comment]
Just checked - zero bulk density issue is not related to presence of wells, it happens in flow-mechanics sequential too.

**Comment by paveltomin** [comment]
> Just checked - zero bulk density issue is not related to presence of wells, it happens in flow-mechanics sequential too.

documented the wrong behavior here https://github.com/GEOS-DEV/GEOS/issues/2698

File: inputFiles/poromechanics/PoroElastic_staircase_co2_3d_sequential.xml
Comment (joshua-white) [inline inputFiles/poromechanics/PoroElastic_staircase_co2_3d_sequential.xml:1]: Is it possible to put some of these blocks in a base XML file to avoid redundancy?

File: inputFiles/poromechanics/PoroElastic_staircase_co2_3d_sequential.xml
Comment (joshua-white) [inline inputFiles/poromechanics/PoroElastic_staircase_co2_3d_sequential.xml:16]: Is this disabled because it doesn't work or just not needed?

File: src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanics.cpp
Comment (joshua-white) [inline src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanics.cpp:77]: Are you inserting `this->` here as a stylistic preference, or is it strictly required here?  I don't know what our code style guidelines say, but we should be consistent.

**Review by joshua-white — COMMENTED** [review]
This looks good to me.  May be worth a quick walkthrough at a dev meeting.

File: inputFiles/poromechanics/PoroElastic_staircase_co2_3d_sequential.xml
Comment (paveltomin) [reply inputFiles/poromechanics/PoroElastic_staircase_co2_3d_sequential.xml:1]: yes i will definitely do that before merging, thanks!

File: inputFiles/poromechanics/PoroElastic_staircase_co2_3d_sequential.xml
Comment (paveltomin) [reply inputFiles/poromechanics/PoroElastic_staircase_co2_3d_sequential.xml:16]: just not needed and also my personal preference about line search

File: src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanics.cpp
Comment (paveltomin) [reply src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanics.cpp:77]: for me it just does not compile without this because of the template introduction
i believe i took that "this" way from some other solver code, see https://github.com/search?q=repo%3AGEOS-DEV%2FGEOS%20this-%3EregisterWrapper&type=code

Comment (joshua-white) [reply src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanics.cpp:77]: ahh, right, definitely required then.

File: inputFiles/poromechanics/PoroElastic_staircase_co2_3d_sequential.xml
Comment (paveltomin) [reply inputFiles/poromechanics/PoroElastic_staircase_co2_3d_sequential.xml:16]: sorry, clarification here, coupled solver currently does not support line search https://github.com/GEOS-DEV/GEOS/blob/5b20fd397c152fadc4e7f12b6209580221295634/src/coreComponents/physicsSolvers/multiphysics/CoupledSolver.hpp#L565

File: inputFiles/poromechanics/PoroElastic_staircase_co2_3d_sequential.xml
Comment (paveltomin) [reply inputFiles/poromechanics/PoroElastic_staircase_co2_3d_sequential.xml:1]: done

File: inputFiles/poromechanics/PoroElastic_staircase_singlephase_3d_seq.xml
Comment (paveltomin) [reply inputFiles/poromechanics/PoroElastic_staircase_singlephase_3d_seq.xml]: done

**Comment by paveltomin** [comment]
> This looks good to me. May be worth a quick walkthrough at a dev meeting.

I think I addressed all your comments, can you have another look and approve if everything looks ok?

**Review by joshua-white — APPROVED** [review]
Looks good.  I think a few more duplicated xml blocks can be moved to the base file, but I could be wrong.  Regardless, I think this is ready to merge.

**Review by CusiniM — APPROVED** [review]
I find it quite hard to remember which solve is which since the names are similar but putting reservoir before or after changes the coupling.  However, I guess that we don't really have a choice atm. It would be good if we could add a check in case someone specified `couplingType="Sequential" ` for a solver `SinglePhasePoromechanicsReservoir`, which, if I am not mistaking, only supports FIM coupling.

**Comment by paveltomin** [comment]
> I find it quite hard to remember which solve is which since the names are similar but putting reservoir before or after changes the coupling. However, I guess that we don't really have a choice atm. It would be good if we could add a check in case someone specified `couplingType="Sequential" ` for a solver `SinglePhasePoromechanicsReservoir`, which, if I am not mistaking, only supports FIM coupling.

I agree it is very confusing. This PR is just to enable the functionality and we can address the details later.
SinglePhasePoromechanicsReservoir does work in sequential way, it just does what is not very expected from user point of view because it is solving reservoir-flow+mechanics followed by solving wells and outer loop on top of that. However it is completely clear behavior from the coupled solver logic.
I suggest we merge it, start using and then straighten the input/code organization.

**Pull request merged**
